libs/sql-parser/src/main/java/io/crate/sql/parser/CaseSensitiveAstBuilder.java

```diff
@@ -0,0 +1,43 @@
+/*
+ * Licensed to Crate.io GmbH ("Crate") under one or more contributor
+ * license agreements.  See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.  Crate licenses
+ * this file to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+package io.crate.sql.parser;
+
+import io.crate.sql.parser.antlr.v4.SqlBaseParser;
+import io.crate.sql.tree.Node;
+import io.crate.sql.tree.StringLiteral;
+
+/**
+ * For Postgres compatibility, it has been the norm that unquoted identifiers are converted to lower cases.
+ * See {@link AstBuilder#visitUnquotedIdentifier(SqlBaseParser.UnquotedIdentifierContext)}
+ * 
+ * However, in some cases the original case sensitivity should be preserved.
+ * For example, a generated reference, "col1 AS MySchema.arr_max(..)" stored in the metadata as a string
```

Comment:
Just to clarify, also for pgsql:
`create schema MySchema` creates a lower case `myschema`
so to be able to use mix case and non-letter chars you always need to use double quotes:
`create schema "MySchema"`

Reply:
Thanks for clarifying. In few cases, we need to regenerate the AST from the sql again (not the original but a sql with double quotes already stripped since it was parsed once)

Reply:
> I do not completely understand this. Shouldn't we rather fix that an identifier is stored in lowercase if unquoted? And add missing quotes?

Could you have a quick look to check if this follows your suggestion? https://github.com/crate/crate/commit/f23ed24ef93ea2ccec88d89d105fb49d8440fae2